### ISSUE TITLE
JSDOC / Comments not preserved?

### ISSUE THREAD

**Issue by Sliov** [nasheomirro/react-polymorphed#6]
Hey there,

Thank you for providing this library, it has been super helpful!

I just noticed that the JSDOC / Typescripts comments are dropped when using the `PolyRefFunction` type, here's a quick example:

With a simple `React.forwardRef`, JSDOC/comments stay intact:

```ts
import * as React from 'react';
import { cva, type VariantProps } from 'class-variance-authority';

export const styles = cva('', {
  variants: {
    /**
     * The intent, or color, of the button.
     * @default primary
     */
    intent: {
      primary: 'bg-black',
    },

    /**
     * The size of the button.
     * @default medium
     */
    size: {
      small: 'p-1',
      medium: 'p-2',
      large: 'p-3',
    },
  },

  defaultVariants: {
    intent: 'primary',
    size: 'medium',
  },
});

type ButtonVariants = VariantProps<typeof styles>;

export type ButtonProps = React.ComponentPropsWithoutRef<'button'> &
  ButtonVariants;

export const Button = React.forwardRef<HTMLButtonElement, ButtonProps>(
  ({ intent, size }, ref) => {
    return (
      <button className={styles({ intent, size })} ref={ref}>
        Click me!
      </button>
    );
  }
);


const Example = () => (
  <div>
    <Button intent={}></Button>
  </div>
);
```
<img width="679" alt="Screenshot 2023-07-03 at 9 49 15 pm" src="https://github.com/nasheomirro/react-polymorphed/assets/121665882/1ae2a9c8-8a74-4994-a456-40e46f53d23e">

But disappears when using the polyRef helper:

```ts
export const polyRef = React.forwardRef as PolyRefFunction;

export const Button = polyRef<'button', ButtonProps>(
  ({ intent, size }, ref) => {
    return (
      <button className={styles({ intent, size })} ref={ref}>
        Click me!
      </button>
    );
  }
);
```

<img width="468" alt="Screenshot 2023-07-03 at 9 49 25 pm" src="https://github.com/nasheomirro/react-polymorphed/assets/121665882/15faf323-5fa7-4349-b193-ca6ef9b0bd13">

Do you know how to fix it or guide through a potential fix? I'm pretty new to TypeScript but happy to dig a bit more to help if pointed in the right direction ✌️ 

Here's the full gist, just in case: https://gist.github.com/Sliov/9966a3fd3ad5c886e6a192a563839284

**Comment by nasheomirro** [comment]
Still figuring out how to fix this, I did find out what the problem is however, something about how the comments are lost when using union types ([playground here](https://www.typescriptlang.org/play?#code/C4TwDgpgBACgTgezAZygXigbwLACgoFQD0AVCVAEYCGAdrVVCUXoVFQFxTLBwCWNAcwDceAL548oSFACqAHgAqAPnRQFUAGRZKnbn0FRRUAD5rN2ipxoBXALYUIcQyNxToAOXnLV6rZh1cPPwCzhK4RERQwAAWvKgAJggQyDQA5MBQ0VQAbtC8wAB0eADGCDTcbJzy8EjIKhj+HFCpqQA0AS2GYRGU1hkxcVCJyUW4peUZllCecjUo9dpNLe1TneK4eFQFVELEkflQpbbJlFTFANZR0QjWAtEFUADKCFBlADYgrzRDEMVvVHAqMBeGUoAdkBAILZkKMKNsXHgerx4pdrgB3Q4IWzHGjAVBohBwS4UPpgjIQqGoN68c7QNHRCDfGLQNxQAQQPFQawQ+Jg74Maw0EFM8B5DLszlvBDcApAA)). 

Now the problem here is that I use a janky trick for correctly inferring props which involves using unions to fix (#2 ). The good news is that that trick isn't needed anymore cause somehow miraculously typescript (i think ts-server) was now able to infer props correctly, the bad news is that in some cases where heavier computations on types are needed it incorrectly infers props again and the trick is again needed to correct it.

I'm unsure whether or not to remove the trick cause I haven't really checked if the problem still persist on heavy computations (last time I saw the problem was like 5 months ago), so I'm a bit lost on whether to remove it now or not.
